### ISSUE TITLE
xrdcp TPC mode operation

### ISSUE THREAD

**Issue by costing** [xrootd/xrootd#270]
Hi,

It seems that when --tpc only is passed to the client it only queries the source to see if it supports TPC but not the target, which causes problems when the target doesn't support TPC in fact, for example dCache storages or older xrootd servers. Would it be possible to ask both endpoints for their capabilities before attempting to transfer anything?

Cheers,

.costin

**Comment by esindril** [comment]
@ljanyst any reason why you removed the check for the target?
https://github.com/xrootd/xrootd/blob/master/src/XrdCl/XrdClThirdPartyCopyJob.cc#L540

**Comment by ljanyst** [comment]
Because it was only possible to test the manager of the target cluster instead of the actual disk server, which is pointless. The disk server can be determined only after the first open succeeded. It is too late to verify whether a TPC is possible after you already assumed it is and are in the middle of the procedure.

@abh3 had very good non-technical reasons for designing the TPC protocol the way it is. However, if we only look at it from technical perspective, it is much much more complicated than it should be. I am not surprised that you encounter problems.

**Comment by esindril** [comment]
Thanks a lot @ljanyst! Ok, so I will tag this one as "Won't fix". @abh3 if there is nothing else to add please close it.

**Comment by abh3** [comment]
Well, there were technical and non-technical issues regarding TPC. One of the main reason the rendezvous approach was taken is because it is authentication agnostic. I did not want to restrict people to particular authentication algorithms (i.e. only those that support forwarding) and I did not want to special case the code for each possible authentication algorithm as this is open-ended. Furthermore, there was no guarantee that the source and destination would use the same authentication algorithm. So, you should start seeing the devil's nest that is quickly appearing just to support some basic level of security. All of that is avoided by using a rendezvous algorithm which, from a practical standpoint, is less complicated than the alternatives if you want to do proper authentication.

Anyway, there is no reason why TPC availability could not be checked at the destination after the file is opened R/W. If support is missing then the copy could be aborted or dropped back at that point. I think the complaint is that when TPC is supported at the source but not the destination you have to wait for a timeout. From a practical standpoint that need not happen.

**Comment by abh3** [comment]
OK, I am closing this.
